masterclass: Initial PR for balenaCLI advanced masterclass

hedss opened the pull request:
Change-type: major
Signed-off-by: Heds Simons <heds@balena.io>

Commit:
masterclass: Initial PR for balenaCLI advanced masterclass

Change-type: major
Signed-off-by: Heds Simons <heds@balena.io>

xginn8 commented on README.md:
```diff
@@ -1 +1,961 @@
+# Prerequisite Classes
+
+This masterclass builds upon knowledge that has been taught in previous classes.
+To gain the most from this masterclass, we reommend that you first undertake
```
```suggestion
To gain the most from this masterclass, we recommend that you first complete
```

hedss replied:
Done

shaunmulligan commented on README.md:
```diff
@@ -1 +1,961 @@
+* A suitable shell environment for command execution (such as `bash`)
+* A [balenaCloud](https://www.balena.io/) account
+* A familiarity with [Dockerfiles](https://docs.docker.com/engine/reference/builder/)
+* A [NodeJS](https://nodejs.org/en/) installation, including [NPM](https://www.npmjs.com/get-npm). The use of [`nvm`](https://github.com/nvm-sh/nvm)
```
I feel like we should rather recommend the installers and avoid the node + Npm install path for the CLI.

hedss replied:
Changed to fall inline with what I've written in the other masterclass.

shaunmulligan commented on README.md:
````diff
@@ -1 +1,961 @@
+  heds/vpn-changed
+```
+Note that the applications listed are not all of the `fincm3` device type, but
+share the common `armv7hf` architecture. Be very careful to select an
````
I think there was a newer update that would allow moving into armv6 fleets too. I think thodoris add the mix capability fleets to the CLI.

hedss replied:
Clarified this.

shaunmulligan commented on README.md:
```diff
@@ -1 +1,961 @@
+catering with different situations, including the ability to use a particular
+release. However, for this exercise, we'll simply use the latest version of
+the code we previously pushed to `cliApp`. In a terminal, execute the following:
+>>>>> NOTE THAT THIS DOESN'T WORK ON MACOS, AS WE SEE AN ERROR WHEN ATTEMPTING
```
Is this error because docker for MacOS doesn't support AUFS?

hedss replied:
I need to look into this in a bit more detail. If we're assuming AUFS, then definitely, but I don't think I've ever managed to get it working under macOS.

hedss replied:
Yeah, I found out why. When Docker did a major update change, it wiped previously shared directories. I've had to re-add my `/Work` directory. Sigh.

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+
+# Introduction
+
+The balena Command Line Interface (balenaCLI) utility consists of a number
```
```suggestion
The balena Command Line Interface (balena-cli) utility consists of a number
```

xginn8 replied:
This change is just for consistency with how we name/call the product (https://github.com/balena-io/balena-cli)

hedss replied:
Hmm, I'd rather change this to 'balena CLI' as how we refer to it in the docs, but not the package name. I know the package name is very similar, but 'balena CLI' refers to the CLI method of operating with balena, rather than the package name (and if we ever changed the package name, it's still valid).

xginn8 replied:
As long as it's consistent, that's fine with me!

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+of commands that allow a user to develop, deploy and manage balena applications
+and devices.
+
+The previous balenaCLI masterclass covered some of the most common techniques
```
ibid.

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+and devices.
+
+The previous balenaCLI masterclass covered some of the most common techniques
+and topics related to balenaCLI. This masterclass aims to build on top of that,
```
ibid.

hedss replied:
Shortened.

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+
+# Hardware and Software Requirements
+
+The hardware and software requirements for the masterclass. This should be a
```
Is this leftover template?

hedss replied:
Very good spot. Yes!

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+to the staging environment using the environment variable.
+
+Token files are saved separately and can be found in the user's home directory
+here: `~/.balena/token`.
```
Is this true in non *nix environments? I've no idea how Windows file hierarchies work.

hedss replied:
For WSL, yes. For normal shell, no idea. I've clarified this.

xginn8 commented on README.md:
````diff
@@ -1 +1,976 @@
+$ echo "balenaUrl: 'balena-staging.com'" > ~/balenaStaging/balenarc.yml
+$ echo "balenaUrl: 'balena-cloud.com'" > ~/balenaProduction/balenarc.yml
+```
+Note that we do *not* prefix the `balenarc.yml` file with a `.` to make it
````
```suggestion
Note that we do *not* prefix the `balenarc.yml` file with a `.` to avoid hiding it
```

xginn8 commented on README.md:
````diff
@@ -1 +1,976 @@
+54323   bbbtest          beaglebone-black 0              0
+```
+
+We'll be using the two separate environments in the next set of exercises
````
@wrboyce has some nice bash aliases that help managing multiple envs with the CLI, perhaps that's worth including?

wrboyce replied:
Not sure they're good for public consumption really ;)

hedss replied:
Yeah, I'd rather stick to actual commands here rather than 'bash magic'. If users want to do scripting that's fine, but we really shouldn't be adding scripts here.

wrboyce replied:
Would be a worthwhile hack friday to add multiple environment support to `balena-cli` mind. I envisage something similar to how aws cli tooling works.

hedss replied:
@wrboyce Dude, that would be utterly awesome

xginn8 commented on README.md:
````diff
@@ -1 +1,976 @@
+To interactively determine which application to move a device to, simply use
+its UUID with the `balena device move` command:
+```
+$ BALENARC_DATA_DIRECTORY=. balena device move 7654321
````
misread this as the device ID, opened https://github.com/balena-io/balena-cli/issues/1457 because that behavior was confusing for me.

hedss replied:
Yeah, UUID is pretty much used for every command... except where it's not. Issue's a good call.

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+## 2.2 Moving Devices between Environments
+
+Moving a device between balena environments is slightly more involved,
+and alters depending on whether you're using a device running a development
```
```suggestion
and differs depending on whether you're using a device running a development
```

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+First, provision a device using a development image. You can do this using
+the balenaCloud dashboard and downloading a 'Development' edition image from
+the `cliApp` application page. Provision your balenaFin with this image
+using either balenaEtcher or balenaCLI.
```
Real question: how can you provision a fin with the CLI? I've never heard of this, and am super curious. Maybe i missed this in the `beginners` masterclass?

xginn8 replied:
ah, ha, farther down (perhaps worth linking to internally)?

hedss replied:
Yes, it's further down, I'll link.

xginn8 commented on README.md:
````diff
@@ -1 +1,976 @@
+As we now have the local IP address for it, we can use this to call the command
+to leave the balenaCloud environment:
+```
+$ BALENARC_DATA_DIRECTORY=. balena leave 192.168.1.171
````
Not important, but I tried this with a NUC VM running via the VBox NAT networking stack and the command just hung.

hedss replied:
Interesting, tried this on both a Pi4 and Fin and it worked. What did the Supervisor on the VM say?

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+
+### 2.2.2 Production Devices
+
+>>> THIS WILL NOT WORK FOR A PRODUCTION DEVICE, BECAUSE AS SOON AS YOU USE
```
maybe worth linking to any relevant issue(s) in all bug sections?

hedss replied:
Yup

hedss replied:
Done

xginn8 commented on README.md:
```diff
@@ -1 +1,976 @@
+
+# 5. Preloading and Preregistering
+
+Whilst for many cases provisoining a device, moving it to an installation
```
```suggestion
Whilst for many cases provisioning a device, moving it to an installation
```

xginn8 reviewed (approved):
looks great! I learned a lot digging into this.

hedss commented:
@xginn8 @shaunmulligan I've updated the PR and rebased. It addresses all the points you raised. I'm in two minds about this. On the one, I'd like to merge, on the other the bug sections are rough and ready and I don't know if we should officially do this. Thoughts?

xginn8 commented:
> @xginn8 @shaunmulligan I've updated the PR and rebased. It addresses all the points you raised. I'm in two minds about this. On the one, I'd like to merge, on the other the bug sections are rough and ready and I don't know if we should officially do this. Thoughts?

As long as the bugs are well-advertised (which they are, you've made sure of that!), then I think the more exposure the better. Maybe not suitable for _everyone_ (since there are obvious known bugs), but still more iterations will give us more + richer data. /my2c

shaunmulligan commented:
@hedss yeah I think lets get it merged and have people start trying it. I agree that if the bugs are documented, it should be fine
